### PR TITLE
Fix auth for dynamic_update.php

### LINKED ISSUE #340: Dynamic_Update.php only giving badauth2
Hello,

Trying to use the dynamic update script with ddclient but no matter what i do i get badauth2 returned. Am I using this wrong or something. Can someone shed some light on how to use this?

Server is apache2 with rewrites for /nic/update etc.
Tried on SSL and non SSL
Tried adding single quotes to username and password
heeelp... ;)

### PR DESCRIPTION
**Pull request by chrostek**
Fixes #340

### CODE REVIEW

File: dynamic_update.php [hunk 1]
```diff
@@ -175,25 +176,18 @@ function valid_ip_address($ip)
...
-                OR perm_items.name = 'zone_content_edit_others'
-        )
-";
-$user = $db->queryRow($user_query);
-if (!$user) {
+$user = $db->queryRow("SELECT users.id, users.password FROM users, perm_templ, perm_templ_items, perm_items WHERE users.username=" . $db->quote($auth_username, 'text') . " AND users.active=1 AND perm_templ.id = users.perm_templ AND perm_templ_items.templ_id = perm_templ.id AND perm_items.id = perm_templ_items.perm_id AND (perm_items.name = 'zone_content_edit_own' OR perm_items.name = 'zone_content_edit_others')");
+
+if ($user) {
+    if (Poweradmin\Password::verify($auth_password, $user['password'])) {
+        if (Poweradmin\Password::needs_rehash($rowObj['password'])) {
```
Comment (edmondas) [inline dynamic_update.php:183]: maybe $user should be here instead of $rowObj?

File: dynamic_update.php [hunk 2]
```diff
@@ -175,25 +176,18 @@ function valid_ip_address($ip)
...
+            Poweradmin\Password::update_user_password($rowObj["id"], $session_pass);
```
Comment (edmondas) [inline dynamic_update.php:184]: not sure if $session_pass is available

Comment (edmondas) [reply dynamic_update.php:184]: $rowObj["id"] -> $user["id"] ?

File: dynamic_update.php [hunk 3]
```diff
@@ -175,25 +176,18 @@ function valid_ip_address($ip)
...
-        AND perm_templ_items.templ_id = perm_templ.id
-        AND perm_items.id = perm_templ_items.perm_id
-        AND (
-                perm_items.name = 'zone_content_edit_own'
-                OR perm_items.name = 'zone_content_edit_others'
-        )
-";
-$user = $db->queryRow($user_query);
-if (!$user) {
+$user = $db->queryRow("SELECT users.id, users.password FROM users, perm_templ, perm_templ_items, perm_items WHERE users.username=" . $db->quote($auth_username, 'text') . " AND users.active=1 AND perm_templ.id = users.perm_templ AND perm_templ_items.templ_id = perm_templ.id AND perm_items.id = perm_templ_items.perm_id AND (perm_items.name = 'zone_content_edit_own' OR perm_items.name = 'zone_content_edit_others')");
```
Comment (edmondas) [inline dynamic_update.php:179]: Maybe that long sql statement can be divided into multiple lines?

**Review by edmondas — CHANGES_REQUESTED** [review]

File: dynamic_update.php (hunk 3 above)
Comment (chrostek) [reply dynamic_update.php]: Ok

File: dynamic_update.php (hunk 2 above)
Comment (chrostek) [reply dynamic_update.php]: removed update_user_pass completely - it is not in Poweradmin\Password and an include of the file in which it is makes another error ...

File: dynamic_update.php (hunk 1 above)
Comment (chrostek) [reply dynamic_update.php]: thanks

File: dynamic_update.php [hunk 4]
```diff
@@ -175,25 +176,22 @@ function valid_ip_address($ip)
...
-        AND perm_items.id = perm_templ_items.perm_id
-        AND (
-                perm_items.name = 'zone_content_edit_own'
-                OR perm_items.name = 'zone_content_edit_others'
-        )
-";
-$user = $db->queryRow($user_query);
-if (!$user) {
+$user = $db->queryRow("SELECT users.id, users.password FROM users, perm_templ, perm_templ_items, perm_items 
+                        WHERE users.username=" . $db->quote($auth_username, 'text') . "
```
Comment (edmondas) [inline dynamic_update.php:180]: Perhaps you could use $username which is defined at line 135 instead? You can also remove lines 136-137.

File: dynamic_update.php [hunk 5]
```diff
@@ -175,25 +176,22 @@ function valid_ip_address($ip)
...
-        AND (
-                perm_items.name = 'zone_content_edit_own'
-                OR perm_items.name = 'zone_content_edit_others'
-        )
-";
-$user = $db->queryRow($user_query);
-if (!$user) {
+$user = $db->queryRow("SELECT users.id, users.password FROM users, perm_templ, perm_templ_items, perm_items 
+                        WHERE users.username=" . $db->quote($auth_username, 'text') . " 
+                        AND users.active=1
```
Comment (edmondas) [inline dynamic_update.php:181]: The equation should be surrounded by spaces

File: dynamic_update.php [hunk 6]
```diff
@@ -175,25 +176,22 @@ function valid_ip_address($ip)
...
+                        AND users.active=1 
+                        AND perm_templ.id = users.perm_templ 
+                        AND perm_templ_items.templ_id = perm_templ.id 
+                        AND perm_items.id = perm_templ_items.perm_id 
+                        AND (
+                            perm_items.name = 'zone_content_edit_own' 
+                            OR perm_items.name = 'zone_content_edit_others'
+                        )");
+
+if ($user) {
```
Comment (edmondas) [inline dynamic_update.php:190]: I think that can be simplified to
```
if (!$user || !Poweradmin\Password::verify($auth_password, $user['password'])) {
    return status_exit('badauth2');
}
```

**Review by edmondas — CHANGES_REQUESTED** [review]

File: dynamic_update.php (hunk 4 above)
Comment (chrostek) [reply dynamic_update.php]: ok

File: dynamic_update.php (hunk 5 above)
Comment (chrostek) [reply dynamic_update.php:179]: copied from 
https://github.com/poweradmin/poweradmin/blob/758c62aa5565dc7ebe41e66f8a62c50b7a3e6ea6/inc/plugins/auth_local/auth_local.plugin.php#L188
;)
but yes, should be surrounded by spaces

File: dynamic_update.php (hunk 6 above)
Comment (chrostek) [reply dynamic_update.php]: ok

**Comment by edmondas** [comment]
@chrostek Thank you for contributing

**Pull request merged**